doc: NCSDK8489 9160 userguide fota samples upd (#4108)
nrfconnect/sdk-nrf: doc/nrf/ug_nrf9160.rst

Introduction
************

The nRF9160 SiP integrates an application MCU, a full LTE modem, an RF front end, and power management.
With built-in GPS support, it is dedicated to asset tracking applications.

> For more details on the SiP, see the `nRF9160 product website`_ and the `nRF9160 product specification`_.

Author: fixed

There are two ways to update the modem firmware:

Full upgrade
  A full upgrade of the modem firmware can be done through either a wired  or a wireless connection.
  When using a wired connection, the upgrade is done through the nRF Connect Programmer, which is part of `nRF Connect for Desktop`_.
  See `Updating the nRF9160 DK cellular modem`_ in the nRF Connect Programmer User Guide for instructions.

>   It is also possible to use the Simple Management Protocol (SMP) to provide an interface over UART which enables the device to do full modem firmware updates.

Reviewer: This is a bit misleading, since nRF Connect for Desktop leverages this protocol as well. The sample provides a python script which use the nrfjprog API directly. This is what the customers would use to see how the update procedure is implemented on the PC side. I'm not sure how to phrase it, but in broad lines:
- Wired update is done using SMP
- nRF Connect for Desktop can be used for performing full modem update regardless of existing firmware
- There is a sample fmfu_smp_svr_sample which shows how to integrade this full modem update over SMP support in an application.
- This sample also provides a python script which shows how to implement the PC side of the update.

The difference between nrf Connect for Desktop and the python script in the sample is that nrf connect for desktop will work regardless of the existing firmware on the device, while the script requires the specific sample to be programmed.

Author: Please check the rephrased section

There are two ways to update the modem firmware:

Full upgrade

>   Both a wired or a wireless connection are suitable to proceed to a full upgrade of the modem firmware:

Author: fixed

>   * When using a wired connection, you can use either the *nRF Connect Programmer*, as part of `nRF Connect for Desktop`_, or a custom application.

Author: fixed

>     Both methods leverage the Simple Management Protocol (SMP) to provide an interface over UART which enables the device to perform the update.

Author: fixed

>     The nRF Connect Programmer can be used to perform the update, regardless of the existing firmware.

Author: The way I interpreted this is that when not using the nRF connect programmer, the discriminating factors that would allow (or not) a user to perform the update would be the kind of images that are part of the existing firmware.

If one of the images contains a custom application similar to the sample linked a bit later in the text, then the update would be possible without the nrf connect programmer. If that would not be the case, only through the nrf connect programmer the user would be able to update the firmware.

@hakonfam is this correct?

>     The nRF Connect Programmer can be used to perform the update, regardless of the existing firmware.

Reviewer: To avoid confusion I will try to state some simple facts here first, then I will answer the questions.

- nRF connect programmer can update the modem firmware regardless of the firmware currently programmed to the device. It loads some code to RAM and executes that AFAICR.
- The `pynrfjprog` python package provides an API for loading modem firmware to a device which is compatible with its "full modem DFU management" management protocol.
- The `lib_fmfu_mgmt` library implements the protocol mentioned above.
- The `fmfu_smp_svr_sample` shows the integration of the `lib_fmfu_mgmt` library.
- Alongside `fmfu_smp_svr_sample` there is a python script which shows the use of the `pynrfjprog`  python package.
- Hence, the python script mentioned above is able to update the modem firmware of a device _given_ that the `fmfu_smp_svr_sample` or any other sample which correctly integrates the `lib_fmfu_mgmt` library is programmed.

> You need some firmware, don't you?

Not when updating the modem firmware using nRF Connect programmer.

> The way I interpreted this is that when not using the nRF connect programmer, the discriminating factors that would allow (or not) a user to perform the update would be the kind of images that are part of the existing firmware.

If by "kind of images" you mean that the image must integrate the `lib_fmfu_mgmt` library, then yes.

> If one of the images contains a custom application similar to the sample linked a bit later in the text, then the update would be possible without the nrf connect programmer. If that's not the case, only through the nrf connect programmer the user would be able to update the firmware. 

Correct.

>     The nRF Connect Programmer can be used to perform the update, regardless of the existing firmware.

Reviewer 2: ok :)

>     The nRF Connect Programmer can be used to perform the update, regardless of the existing firmware.

Author: Updated.

    See `Updating the nRF9160 DK cellular modem`_ in the nRF Connect Programmer user guide for more details.

>     See the :ref:`fmfu_smp_svr_sample` for a sample meant to showcase the integration of the :ref:`device_mgmt` subsystem in a custom application.

Reviewer: ```suggestion
    See the :ref:`fmfu_smp_svr_sample` for a sample meant to showcase the integration of the :ref:`lib_fmfu_mgmt` subsystem in a custom application.
```

Reviewer: We should point to the mgmt module for full modem update specifically

Author: Fixed

>     See the :ref:`fmfu_smp_svr_sample` for a sample meant to showcase the integration of the :ref:`lib_fmfu_mgmt` subsystem in a custom application.

Reviewer 2: ```suggestion
    See the :ref:`fmfu_smp_svr_sample` sample for an example on how to integrate the :ref:`lib_fmfu_mgmt` subsystem in a custom application.
```

Author: fixed

>   * When using a wireless connection, the upgrade is applied as firmware over-the-air (FOTA).

Reviewer 2: "is applied over the air", or "is applied as firmware over-the-air (FOTA) upgrade."

Author: fixed

     You can hardcode the information in the application, or you can use functionality like AWS jobs to provide the URL dynamically.

The full FOTA procedure depends on where the binary files are hosted for download.

You can refer to the following implementation samples:

* :ref:`http_modem_delta_update_sample` - performs a full firmware OTA update of the modem.

> * :ref:`_http_full_modem_update_sample` - performs a delta OTA update of the modem firmware.

Reviewer: This and the delta sample line has switched the descriptions

Author: Fixed

To program the HEX file, use nrfjprog (which is part of the `nRF Command Line Tools`_).

If you want to route some pins differently from what is done in the preprogrammed firmware, program the :ref:`zephyr:hello_world` sample instead of the preprogrammed firmware.
Build the sample (located under ``ncs/zephyr/samples/hello_world``) for the nrf9160dk_nrf52840 development kit.

> To change the routing options, enable or disable the corresponding devicetree nodes for that development kit as needed.

Reviewer 3: nrf9160dk_nrf52840 is not a development kit and such term should not be used in this context. According to https://github.com/nrfconnect/sdk-nrf/pull/3708, it is a board. Perhaps "target" could be also used here, as this part refers to building.

Reviewer 2: Correct, this should be board.

Author: fixed

See :ref:`zephyr:nrf9160dk_board_controller_firmware` for detailed information.

> DK revisions

Reviewer 2: This should be Board revisions.

Author: fixed, including the following the paragraph
